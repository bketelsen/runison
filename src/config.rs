use figment::{
    providers::{Format, Toml},
    Error, Figment,
};
use serde::{Deserialize, Serialize};

#[derive(Clone, PartialEq, Deserialize)]
pub struct Config {
    pub root: Root,
    pub path: Path,
    pub ignore: Ignore,
}

#[derive(Clone, PartialEq, Deserialize)]
pub struct Root {
    pub path: String,
}
#[derive(Clone, PartialEq, Deserialize)]

pub struct Path {
    pub directories: Vec<String>,
}

#[derive(Clone, PartialEq, Deserialize)]
pub struct Ignore {
    pub name: Vec<String>,
    pub path: Vec<String>,
}

pub fn get_config(path: &str) -> Result<Config, figment::Error> {
    Figment::new().merge(Toml::file(path)).extract()
}
