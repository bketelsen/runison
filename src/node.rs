use crate::config::Config;
use serde::{Deserialize, Serialize};
use std::ffi::OsString;
use std::os::unix::fs::MetadataExt;
use std::path::PathBuf;
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]

pub struct Node {
    pub is_dir: bool,
    pub is_file: bool,
    pub is_symlink: bool,
    pub name: OsString,
    pub path: String,
    pub len: u64,
    pub modified: std::time::SystemTime,
    pub inode: u64,
}
impl Node {
    pub fn new(
        is_dir: bool,
        is_file: bool,
        is_symlink: bool,
        name: OsString,
        path: String,
        len: u64,
        modified: std::time::SystemTime,
        inode: u64,
    ) -> Option<Node> {
        Some(Node {
            is_dir,
            is_file,
            is_symlink,
            name,
            path,
            len,
            modified,
            inode,
        })
    }
    pub fn from_path(path: &str, config: &Config) -> Option<Node> {
        // add the root path back to the given path to get the full file path
        let config = config.clone();
        let rp = String::from(config.root.path.clone());
        let mut pathbuf = PathBuf::from(&rp);
        pathbuf.push(path);

        // get the metadata of the file
        // TODO: handle error for bad symlinks
        let metadata = std::fs::metadata(&pathbuf).unwrap();
        let inode = metadata.ino();
        let filetype = metadata.file_type();

        let root = Node {
            is_dir: filetype.is_dir(),
            is_file: filetype.is_file(),
            is_symlink: filetype.is_symlink(),
            name: PathBuf::from(path).file_name().unwrap().to_os_string(),
            path: pathbuf.to_str().unwrap().to_string(),
            len: metadata.len(),
            modified: metadata.modified().unwrap(),
            inode,
        };
        Some(root)
    }
}
