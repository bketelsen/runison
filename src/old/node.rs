use crate::config::{Config, Path};
use serde::{Deserialize, Serialize};
use std::ffi::OsString;
use std::os::unix::fs::MetadataExt;
use std::path::PathBuf;
#[derive(Clone, PartialEq, Serialize, Deserialize)]

pub struct Node {
    pub is_dir: bool,
    pub is_file: bool,
    pub is_symlink: bool,
    pub name: OsString,
    pub path: PathBuf,
    pub relative_path: PathBuf,
    pub len: u64,
    pub modified: std::time::SystemTime,
    pub inode: u64,
    pub root_path: PathBuf,
}
impl Node {
    pub fn new(
        is_dir: bool,
        is_file: bool,
        is_symlink: bool,
        name: OsString,
        path: PathBuf,
        relative_path: PathBuf,
        len: u64,
        modified: std::time::SystemTime,
        inode: u64,
        root_path: PathBuf,
    ) -> Option<Node> {
        Some(Node {
            is_dir,
            is_file,
            is_symlink,
            name,
            path,
            relative_path,
            len,
            modified,
            inode,
            root_path,
        })
    }
    pub fn from_path(root_path: PathBuf, path: PathBuf, config: &Config) -> Option<Node> {
        // add the root path back to the given path to get the full file path
        let config = config.clone();
        // get the metadata of the file
        // TODO: handle error for bad symlinks
        let mut joined = PathBuf::new();
        for p in root_path.iter() {
            joined.push(p);
        }

        if path.to_str() != Some(".") {
            for p in path.iter() {
                joined.push(p);
            }
        }
        let metadata = std::fs::metadata(&joined).unwrap();
        let inode = metadata.ino();
        let filetype = metadata.file_type();

        let root = Node {
            is_dir: filetype.is_dir(),
            is_file: filetype.is_file(),
            is_symlink: filetype.is_symlink(),
            name: path.clone().into_os_string(),
            path: joined,
            relative_path: path,
            len: metadata.len(),
            modified: metadata.modified().unwrap(),
            inode,
            root_path,
        };
        Some(root)
    }
}
