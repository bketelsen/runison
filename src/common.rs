use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use std::net::SocketAddr;

use crate::config::Config;
use crate::node::Node;

use crate::synchronizer::{Change, Status};

#[derive(Serialize, Deserialize)]
pub enum Message {
    // To DiscoveryServer
    RegisterParticipant(String, SocketAddr),
    UnregisterParticipant(String),
    GetStatus(),
    GetNodes(),
    GetChangeset(BTreeMap<String, Node>),

    // From DiscoveryServer
    ParticipantList(Vec<(String, SocketAddr)>),
    ParticipantNotificationAdded(String, SocketAddr),
    ParticipantNotificationRemoved(String),
    ServerStatus(Status),
    NodeList(BTreeMap<String, Node>),

    Changeset(Vec<Change>),

    // From Participant to Participant
    Greetings(String, String), //name and grettings
}
