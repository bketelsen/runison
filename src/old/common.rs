use std::collections::BTreeMap;

use message_io::network::Endpoint;
use serde::{Deserialize, Serialize};
use std::net::SocketAddr;

use crate::node::Node;
use std::fs::File;

use crate::synchronizer::{Change, Status};

#[derive(Serialize, Deserialize)]
pub enum Message {
    // To DiscoveryServer
    RegisterParticipant(String, SocketAddr),
    UnregisterParticipant(String),
    GetStatus(),
    GetNodes(),
    GetChangeset(BTreeMap<String, Node>),

    // From DiscoveryServer
    ParticipantList(Vec<(String, SocketAddr)>),
    ParticipantNotificationAdded(String, SocketAddr),
    ParticipantNotificationRemoved(String),
    ServerStatus(Status),
    NodeList(BTreeMap<String, Node>),

    Changeset(Vec<Change>),

    // File Transfer
    FileRequest(String, usize), // name, size

    //From sender to receiver
    Chunk(Vec<u8>), // data

    //From receiver to sender
    CanReceive(bool),
    SendMe(String), // name, size

    // From Participant to Participant
    Greetings(String, String), //name and grettings
}
pub struct Transfer {
    pub endpoint: Endpoint,
    pub file: File,
    pub name: String,
    pub current_size: usize,
    pub expected_size: usize,
}
