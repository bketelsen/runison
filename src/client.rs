use crate::config::Config;
use crate::node::Node;
use message_io::events::EventQueue;
use message_io::network::{Endpoint, NetEvent, Network};
use serde::{Deserialize, Serialize};
use std::io;
use std::net::SocketAddr;

use crate::common::Message;

pub enum Event {
    Network(NetEvent<Message>),
}

// Client encapsulates the network activity
pub struct Client {
    pub event_queue: EventQueue<Event>,
    pub network: Network,
    listen: String,
    port: String,
    listen_addr: String,
}
impl Client {
    pub fn new(listen: &str, port: &str) -> Option<Client> {
        let mut event_queue = EventQueue::new();

        let network_sender = event_queue.sender().clone();
        let network = Network::new(move |net_event| network_sender.send(Event::Network(net_event)));

        let listen_addr = format!("{}:{}", &listen, &port);
        Some(Client {
            event_queue,
            network,
            listen: listen.to_string(),
            port: port.to_string(),
            listen_addr,
        })
    }
    // Server
    pub fn start(&mut self) -> io::Result<(usize, SocketAddr)> {
        self.network.listen_tcp(&self.listen_addr)
    }
    // Client
    pub fn connect(&mut self, target: &str) -> io::Result<Endpoint> {
        self.network.connect_tcp(target)
    }
    // Client Gossip
    pub fn gossip(&mut self) -> io::Result<(usize, SocketAddr)> {
        self.network.listen_udp(&self.listen_addr)
    }
}
