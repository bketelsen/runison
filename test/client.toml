[root]
path = "/Users/bjk/src/github.com/bketelsen"

[path]
directories = [
    "src",
    "dotfiles"
]
[ignore]
name = [
    "*~",
    ".*~"
]
path = [
    "*node_modules*"
]