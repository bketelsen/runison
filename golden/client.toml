[root]
path = "/Users/bjk/src/github.com/bketelsen/runison/test/root2"

[path]
directories = [
    "src",
    "dotfiles"
]
[ignore]
name = [
    "*~",
    ".*~"
]
path = [
    "*node_modules*"
]